### ISSUE TITLE
nested sort in perl-5.7.0 fails

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4049]
Migrated from [rt.perl.org#7063](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=7063) (status was 'resolved')

Searchable as RT7063$

**Comment by p5pRT** [comment]
### From js@delphi.kpnqwest.fi



Code&#8203;:

&nbsp;    \#\!/usr/bin/perl \-w

&nbsp;    use strict;
&nbsp;    my @&#8203;x = sort \{ X&#8203;::foo\($a\) cmp X&#8203;::foo\($b\) \} qw\(FOO BAR\);

&nbsp;    package X;
&nbsp;    sub foo \{ 
&nbsp;        my $arg = shift; 
&nbsp;        my @&#8203;dummy = sort \{ $arg\.$a cmp $arg\.$b \} qw\(B A\); 
&nbsp;        print "$arg @&#8203;dummy\\n"; 
&nbsp;        return $arg; 
&nbsp;    \}

Expected output is\, AFAICS\, that kind of lines \(in some order\)&#8203;:

&nbsp;    FOO A B
&nbsp;    BAR A B

However\, this is what I got in my perl\-5\.7\.0&#8203;:

&nbsp;    FOO A B
&nbsp;    Use of uninitialized value in concatenation \(\.\) or string at \./foo\.pl line 9\.
&nbsp;    Use of uninitialized value in concatenation \(\.\) or string at \./foo\.pl line 9\.
&nbsp;    BAR B A

and I've heard that perl\-5\.7\.1 outputs&#8203;:

&nbsp;    FOO A B
&nbsp;    BAR B A

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.7.0:

Configured by js at Fri Jan 26 20:09:22 EET 2001.

Summary of my perl5 (revision 5.0 version 7 subversion 0) configuration:
  Platform:
    osname=solaris, osvers=2.7, archname=sun4-solaris
    uname='sunos delphi 5.7 generic_106541-10 sun4u sparc sunw,ultra-250 '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=undef usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-I/usr/local/include', optimize='-xO4', cppflags='-I/usr/local/include'
    ccversion='WorkShop Compilers 5.0 98/12/15 C 5.0', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=8, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib -L/opt/SUNWspro/SC5.0/lib'
    libpth=/usr/local/lib /opt/SUNWspro/SC5.0/lib /lib /usr/lib /usr/ccs/lib
    libs=-lsocket -lnsl -ldl -lm -lc -lcrypt -lsec
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-KPIC', lddlflags='-G -L/usr/local/lib -L/opt/SUNWspro/SC5.0/lib'

Locally applied patches:
    


@INC for perl v5.7.0:
    /home/js
    /usr/local/perl-5.7.0/lib/5.7.0/sun4-solaris
    /usr/local/perl-5.7.0/lib/5.7.0
    /usr/local/perl-5.7.0/lib/site_perl/5.7.0/sun4-solaris
    /usr/local/perl-5.7.0/lib/site_perl/5.7.0
    /usr/local/perl-5.7.0/lib/site_perl
    .


Environment for perl v5.7.0:
    HOME=/home/js
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/oracle/app/product/8.0.6/lib:/usr/openwin/lib
    LOGDIR (unset)
    PATH=/home/js/bin:/oracle/app/product/8.0.6/bin:/opt/SUNWspro/bin:/bin:/usr/bin:/sbin:/usr/sbin:/usr/ccs/bin:/usr/local/bin:/usr/local/sbin:/usr/X/bin
    PERL5LIB=/home/js
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

still exists in bleeperl@&#8203;19435&#8203;:

\[19435\]$ \./perl \-Ilib /tmp/p
FOO A B
Use of uninitialized value in concatenation \(\.\) or string at /tmp/p line 13\.
Use of uninitialized value in concatenation \(\.\) or string at /tmp/p line 13\.
BAR B A

**Comment by p5pRT** [comment]
### From @smpeters

> \[davem \- Thu May 08 17&#8203;:36&#8203;:31 2003\]&#8203;:
> 
> still exists in bleeperl@&#8203;19435&#8203;:
> 
> \[19435\]$ \./perl \-Ilib /tmp/p
> FOO A B
> Use of uninitialized value in concatenation \(\.\) or string at /tmp/p 
line 13\.
> Use of uninitialized value in concatenation \(\.\) or string at /tmp/p 
line 13\.
> BAR B A
> 

Appears to be fixed with patch included with 36430\.

steve@&#8203;kirk&#8203;:\~/perl\-current$ \./perl \-Ilib rt\_7063\.pl
FOO A B
BAR A B

**Comment by p5pRT** [comment]
### From @smpeters

> \[stmpeters \- Wed Dec 14 11&#8203;:55&#8203;:23 2005\]&#8203;:
> 
> > \[davem \- Thu May 08 17&#8203;:36&#8203;:31 2003\]&#8203;:
> > 
> > still exists in bleeperl@&#8203;19435&#8203;:
> > 
> > \[19435\]$ \./perl \-Ilib /tmp/p
> > FOO A B
> > Use of uninitialized value in concatenation \(\.\) or string at /tmp/p 
> line 13\.
> > Use of uninitialized value in concatenation \(\.\) or string at /tmp/p 
> line 13\.
> > BAR B A
> > 
> 
> Appears to be fixed with patch included with 36430\.
> 
> steve@&#8203;kirk&#8203;:\~/perl\-current$ \./perl \-Ilib rt\_7063\.pl
> FOO A B
> BAR A B
> 

Actually fixed with change \#25953\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'
